README
ONLINE DOCUMENTATION

You can find the latest Spark documentation, including a programming guide,
on the project wiki at http://github.com/mesos/spark/wiki. This file only
contains basic setup instructions.



BUILDING

Spark requires Scala 2.8. This version has been tested with 2.8.1.final.

The project is built using Simple Build Tool (SBT), which is packaged with it.
To build Spark and its example programs, run sbt/sbt update compile.

To run Spark, you will need to have Scala's bin in your $PATH, or you
will need to set the SCALA_HOME environment variable to point to where
you've installed Scala. Scala must be accessible through one of these
methods on Mesos slave nodes as well as on the master.

To run one of the examples, use ./run <class> <params>. For example,
./run spark.examples.SparkLR will run the Logistic Regression example.
Each of the example programs prints usage help if no params are given.

All of the Spark samples take a <host> parameter that is the Mesos master
to connect to. This can be a Mesos URL, or "local" to run locally with one
thread, or "local[N]" to run locally with N threads.



CONFIGURATION

Spark can be configured through two files: conf/java-opts and conf/spark-env.sh.

In java-opts, you can add flags to be passed to the JVM when running Spark.

In spark-env.sh, you can set any environment variables you wish to be available
when running Spark programs, such as PATH, SCALA_HOME, etc. There are also
several Spark-specific variables you can set:
- SPARK_CLASSPATH: Extra entries to be added to the classpath, separated by ":".
- SPARK_MEM: Memory for Spark to use, in the format used by java's -Xmx option
             (for example, 200m meams 200 MB, 1g means 1 GB, etc).
- SPARK_LIBRARY_PATH: Extra entries to add to java.library.path for locating
                      shared libraries.
- SPARK_JAVA_OPTS: Extra options to pass to JVM.

Note that spark-env.sh must be a shell script (it must be executable and start
with a #! header to specify the shell to use).


CHANGES.txt
Spark Change Log
----------------

Release 1.6.1

  [SPARK-13474][PROJECT INFRA] Update packaging scripts to push artifacts to home.apache.org
  Josh Rosen <joshrosen@databricks.com>
  2016-02-26 18:40:00 -0800
  Commit: 8a43c3b, github.com/apache/spark/pull/11350

  [SPARK-13454][SQL] Allow users to drop a table with a name starting with an underscore.
  Yin Huai <yhuai@databricks.com>
  2016-02-26 12:34:03 -0800
  Commit: a57f87e, github.com/apache/spark/pull/11349

  [SPARK-12874][ML] ML StringIndexer does not protect itself from column name duplication
  Yu ISHIKAWA <yuu.ishikawa@gmail.com>
  2016-02-25 13:21:33 -0800
  Commit: abe8f99, github.com/apache/spark/pull/11370

  Revert "[SPARK-13444][MLLIB] QuantileDiscretizer chooses bad splits on large DataFrames"
  Xiangrui Meng <meng@databricks.com>
  2016-02-25 12:28:03 -0800
  Commit: d59a08f

  [SPARK-12316] Wait a minutes to avoid cycle calling.
  huangzhaowei <carlmartinmax@gmail.com>
  2016-02-25 09:14:19 -0600
  Commit: 5f7440b2, github.com/apache/spark/pull/10475

  [SPARK-13439][MESOS] Document that spark.mesos.uris is comma-separated
  Michael Gummelt <mgummelt@mesosphere.io>
  2016-02-25 13:32:09 +0000
  Commit: e3802a7, github.com/apache/spark/pull/11311

  [SPARK-13441][YARN] Fix NPE in yarn Client.createConfArchive method
  Terence Yim <terence@cask.co>
  2016-02-25 13:29:30 +0000
  Commit: 1f03163, github.com/apache/spark/pull/11337

  [SPARK-13444][MLLIB] QuantileDiscretizer chooses bad splits on large DataFrames
  Oliver Pierson <ocp@gatech.edu>, Oliver Pierson <opierson@umd.edu>
  2016-02-25 13:24:46 +0000
  Commit: cb869a1, github.com/apache/spark/pull/11319

  [SPARK-13473][SQL] Don't push predicate through project with nondeterministic field(s)
  Cheng Lian <lian@databricks.com>
  2016-02-25 20:43:03 +0800
  Commit: 3cc938a, github.com/apache/spark/pull/11348

  [SPARK-13482][MINOR][CONFIGURATION] Make consistency of the configuraiton named in TransportConf.
  huangzhaowei <carlmartinmax@gmail.com>
  2016-02-24 23:52:17 -0800
  Commit: 8975996, github.com/apache/spark/pull/11360

  [SPARK-13475][TESTS][SQL] HiveCompatibilitySuite should still run in PR builder even if a PR only changes sql/core
  Yin Huai <yhuai@databricks.com>
  2016-02-24 13:34:53 -0800
  Commit: fe71cab, github.com/apache/spark/pull/11351

  [SPARK-13390][SQL][BRANCH-1.6] Fix the issue that Iterator.map().toSeq is not Serializable
  Shixiong Zhu <shixiong@databricks.com>
  2016-02-24 13:35:36 +0000
  Commit: 06f4fce, github.com/apache/spark/pull/11334

  [SPARK-13410][SQL] Support unionAll for DataFrames with UDT columns.
  Franklyn D'souza <franklynd@gmail.com>
  2016-02-23 15:34:04 -0800
  Commit: 573a2c9, github.com/apache/spark/pull/11333

  [SPARK-13355][MLLIB] replace GraphImpl.fromExistingRDDs by Graph.apply
  Xiangrui Meng <meng@databricks.com>
  2016-02-22 23:54:21 -0800
  Commit: 0784e02, github.com/apache/spark/pull/11226

  [SPARK-12746][ML] ArrayType(_, true) should also accept ArrayType(_, false) fix for branch-1.6
  Earthson Lu <Earthson.Lu@gmail.com>
  2016-02-22 23:40:36 -0800
  Commit: d31854d, github.com/apache/spark/pull/11237

  Preparing development version 1.6.1-SNAPSHOT
  Patrick Wendell <pwendell@gmail.com>
  2016-02-22 18:30:30 -0800
  Commit: 2902798

  Preparing Spark release v1.6.1-rc1
  Patrick Wendell <pwendell@gmail.com>
  2016-02-22 18:30:24 -0800
  Commit: 152252f

  Update branch-1.6 for 1.6.1 release
  Michael Armbrust <michael@databricks.com>
  2016-02-22 18:25:48 -0800
  Commit: 40d11d0

  [SPARK-11624][SPARK-11972][SQL] fix commands that need hive to exec
  Daoyuan Wang <daoyuan.wang@intel.com>
  2016-02-22 18:13:32 -0800
  Commit: f7898f9, github.com/apache/spark/pull/9589

  [SPARK-13298][CORE][UI] Escape "label" to avoid DAG being broken by some special character
  Shixiong Zhu <shixiong@databricks.com>
  2016-02-22 17:42:30 -0800
  Commit: 85e6a22, github.com/apache/spark/pull/11309

  [SPARK-12546][SQL] Change default number of open parquet files
  Michael Armbrust <michael@databricks.com>
  2016-02-22 15:27:29 -0800
  Commit: 699644c, github.com/apache/spark/pull/11308

  [SPARK-13371][CORE][STRING] TaskSetManager.dequeueSpeculativeTask compares Option and String directly.
  Sean Owen <sowen@cloudera.com>
  2016-02-18 12:14:30 -0800
  Commit: 16f35c4, github.com/apache/spark/pull/11253

  [SPARK-13350][DOCS] Config doc updated to state that PYSPARK_PYTHON's default is "python2.7"
  Christopher C. Aycock <chris@chrisaycock.com>
  2016-02-17 11:24:18 -0800
  Commit: 66106a6, github.com/apache/spark/pull/11239

  [SPARK-13279] Remove O(n^2) operation from scheduler.
  Sital Kedia <skedia@fb.com>
  2016-02-16 22:27:34 -0800
  Commit: 98354ca, github.com/apache/spark/pull/11175

  Correct SparseVector.parse documentation
  Miles Yucht <miles@databricks.com>
  2016-02-16 13:01:21 +0000
  Commit: d950891, github.com/apache/spark/pull/11213

  [SPARK-13312][MLLIB] Update java train-validation-split example in ml-guide
  JeremyNixon <jnixon2@gmail.com>
  2016-02-15 09:25:13 +0000
  Commit: 71f53ed, github.com/apache/spark/pull/11199

  [SPARK-13300][DOCUMENTATION] Added pygments.rb dependancy
  Amit Dev <amitdev@gmail.com>
  2016-02-14 11:41:27 +0000
  Commit: ec40c5a, github.com/apache/spark/pull/11180

  [SPARK-12363][MLLIB] Remove setRun and fix PowerIterationClustering failed test
  Liang-Chi Hsieh <viirya@gmail.com>, Xiangrui Meng <meng@databricks.com>
  2016-02-13 15:56:20 -0800
  Commit: 107290c, github.com/apache/spark/pull/10539

  [SPARK-13142][WEB UI] Problem accessing Web UI /logPage/ on Microsoft Windows
  markpavey <mark.pavey@thefilter.com>
  2016-02-13 08:39:43 +0000
  Commit: 93a55f3, github.com/apache/spark/pull/11135

  [SPARK-13153][PYSPARK] ML persistence failed when handle no default value parameter
  Tommy YU <tummyyu@163.com>
  2016-02-11 18:38:49 -0800
  Commit: 18661a2, github.com/apache/spark/pull/11043

  [SPARK-13047][PYSPARK][ML] Pyspark Params.hasParam should not throw an error

www/app.py
from datetime import datetime, timedelta
import dateutil.parser
import json
import logging

from flask import Flask, url_for, Markup, Blueprint, redirect, flash
from flask.ext.admin import Admin, BaseView, expose, AdminIndexView
from flask.ext.admin.contrib.sqla import ModelView
from flask import request
from wtforms import Form, DateTimeField

from pygments import highlight
from pygments.lexers import PythonLexer
from pygments.formatters import HtmlFormatter


import markdown
import chartkick

from core.settings import Session
from core import models
from core.models import State
from core import settings
from core import utils

dagbag = models.DagBag(settings.DAGS_FOLDER)
session = Session()

app = Flask(__name__)
app.secret_key = 'fluxified'

# Init for chartkick, the python wrapper for highcharts
ck = Blueprint(
    'ck_page', __name__,
    static_folder=chartkick.js(), static_url_path='/static')
app.register_blueprint(ck, url_prefix='/ck')
app.jinja_env.add_extension("chartkick.ext.charts")


# Date filter form needed for gantt and graph view
class DateTimeForm(Form):
    execution_date = DateTimeField("Execution date")


@app.route('/')
def index():
    return redirect(url_for('admin.index'))


class HomeView(AdminIndexView):
    """
    Basic home view, just showing the README.md file
    """
    @expose("/")
    def index(self):
        md = "".join(
            open(settings.BASE_FOLDER + '/README.md', 'r').readlines())
        content = Markup(markdown.markdown(md))
        return self.render('admin/index.html', content=content)
admin = Admin(app, name="Flux", index_view=HomeView(name='Home'))


class Flux(BaseView):

    def is_visible(self):
        return False

    @expose('/')
    def index(self):
        return self.render('admin/dags.html')

    @expose('/code')
    def code(self):
        dag_id = request.args.get('dag_id')
        dag = dagbag.dags[dag_id]
        code = "".join(open(dag.filepath, 'r').readlines())
        title = dag.filepath.replace(settings.BASE_FOLDER + '/dags/', '')
        html_code = highlight(
            code, PythonLexer(), HtmlFormatter(noclasses=True))
        return self.render(
            'admin/code.html', html_code=html_code, dag=dag, title=title)

    @expose('/log')
    def log(self):
        dag_id = request.args.get('dag_id')
        task_id = request.args.get('task_id')
        execution_date = request.args.get('execution_date')
        dag = dagbag.dags[dag_id]
        loc = settings.BASE_LOG_FOLDER + "/{dag_id}/{task_id}/{execution_date}"
        loc = loc.format(**locals())
        try:
            f = open(loc)
            log = "".join(f.readlines())
        except:
            log = "Log file is missing"

        log = "<pre><code>" + log + "</code></pre>"
        title = "Logs for {task_id} on {execution_date}".format(**locals())
        html_code = log

        return self.render(
            'admin/code.html', html_code=html_code, dag=dag, title=title)

    @expose('/tree')
    def tree(self):
        dag_id = request.args.get('dag_id')
        dag = dagbag.dags[dag_id]
        action = request.args.get('action')

        base_date = request.args.get('base_date')
        if not base_date:
            base_date = datetime.now()
        else:
            base_date = dateutil.parser.parse(base_date)

        num_runs = request.args.get('num_runs')
        if not num_runs:
            num_runs = 45
        else:
            num_runs = int(num_runs)
        from_date = (base_date-(num_runs * dag.schedule_interval)).date()
        from_date = datetime.combine(from_date, datetime.min.time())

        if action == 'clear':
            task_id = request.args.get('task_id')
            task = dag.get_task(task_id)
            execution_date = request.args.get('execution_date')
            future = request.args.get('future') == "true"
            past = request.args.get('past') == "true"
            upstream = request.args.get('upstream') == "true"
            downstream = request.args.get('downstream') == "true"

            start_date = execution_date
            end_date = execution_date

            if future:
                end_date = None
            if past:
                start_date = None

            count = task.clear(
                start_date=start_date,
                end_date=end_date,
                upstream=upstream,
                downstream=downstream)

            flash("{0} task instances have been cleared".format(count))
            return redirect(url_for('flux.tree', dag_id=dag_id))

        dates = utils.date_range(
            from_date, base_date, dag.schedule_interval)
        task_instances = {}
        for ti in dag.get_task_instances(from_date):
            task_instances[(ti.task_id, ti.execution_date)] = ti

        def recurse_nodes(task):
            return {
                'name': task.task_id,
                'instances': [
                    utils.alchemy_to_dict(
                        task_instances.get((task.task_id, d))) or {
                            'execution_date': d.isoformat(),
                            'task_id': task.task_id
                        }
                    for d in dates],
                'children': [recurse_nodes(t) for t in task.upstream_list],
                'num_dep': len(task.upstream_list),
                'operator': task.task_type,
                'retries': task.retries,
                'owner': task.owner,
                'start_date': task.start_date,
                'end_date': task.end_date,
                'depends_on_past': task.depends_on_past,
            }
        if len(dag.roots) > 1:
            # d3 likes a single root
            data = {
                'name': 'root',
                'instances': [],
                'children': [recurse_nodes(t) for t in dag.roots]
            }
        else:
            data = recurse_nodes(dag.roots[0])

        data = json.dumps(data, indent=4, default=utils.json_ser)

        return self.render(
            'admin/tree.html',
            dag=dag, data=data)

    @expose('/graph')
    def graph(self):
        session = settings.Session()
        dag_id = request.args.get('dag_id')
        dag = dagbag.dags[dag_id]

        nodes = []
        edges = []
        for task in dag.tasks:
            nodes.append({
                'id': task.task_id,
                'value': {'label': task.task_id}
            })

        def get_downstream(task, edges):
            for t in task.upstream_list:
                edges.append({
                    'u': t.task_id,
                    'v': task.task_id,
                    'value': {'label': ''}
                })
                get_downstream(t, edges)

        for t in dag.roots:
            get_downstream(t, edges)

        dttm = request.args.get('execution_date')
        if dttm:
            dttm = dateutil.parser.parse(dttm)
        else:
            dttm = dag.latest_execution_date or datetime.now().date()

        form = DateTimeForm(data={'execution_date': dttm})

        task_instances = {
            ti.task_id: utils.alchemy_to_dict(ti)
            for ti in dag.get_task_instances(dttm, dttm)
        }
        tasks = {
            t.task_id: utils.alchemy_to_dict(t)
            for t in dag.tasks
        }
        session.commit()
        session.close()

        return self.render(
            'admin/graph.html',
            dag=dag,
            form=form,
            execution_date=dttm.isoformat(),
            task_instances=json.dumps(task_instances, indent=2),
            tasks=json.dumps(tasks, indent=2),
            nodes=json.dumps(nodes, indent=2),
            edges=json.dumps(edges, indent=2),)

    @expose('/duration')
    def duration(self):
        session = settings.Session()
        dag_id = request.args.get('dag_id')
        dag = dagbag.dags[dag_id]
        from_date = (datetime.today()-timedelta(30)).date()
        from_date = datetime.combine(from_date, datetime.min.time())

        all_data = []
        for task in dag.tasks:
            data = []
            for ti in task.get_task_instances(from_date):
                if ti.end_date:
                    data.append([
                        ti.execution_date.isoformat(),
                        int(ti.duration)
                    ])
            if data:
                all_data.append({'data': data, 'name': task.task_id})

        session.commit()
        session.close()

        return self.render(
            'admin/chart.html',
            dag=dag,
            data=all_data,
            height="500px",
        )

    @expose('/landing_times')
    def landing_times(self):
        session = settings.Session()
        dag_id = request.args.get('dag_id')
        dag = dagbag.dags[dag_id]
        from_date = (datetime.today()-timedelta(30)).date()
        from_date = datetime.combine(from_date, datetime.min.time())

        all_data = []
        for task in dag.tasks:
            data = []
            for ti in task.get_task_instances(from_date):
                if ti.end_date:
                    data.append([
                        ti.execution_date.isoformat(), (
                            ti.end_date - ti.execution_date
                        ).total_seconds()/(60*60)
                    ])
            all_data.append({'data': data, 'name': task.task_id})

        session.commit()
        session.close()

        return self.render(
            'admin/chart.html',
            dag=dag,
            data=all_data,
            height="500px",
        )

    @expose('/gantt')
    def gantt(self):

        session = settings.Session()
        dag_id = request.args.get('dag_id')
        dag = dagbag.dags[dag_id]

        dttm = request.args.get('execution_date')
        if dttm:
            dttm = dateutil.parser.parse(dttm)
        else:
            dttm = dag.latest_execution_date

        form = DateTimeForm(data={'execution_date': dttm})

        tis = dag.get_task_instances(dttm, dttm)
        tis = sorted(tis, key=lambda ti: ti.start_date)
        tasks = []
        data = []
        for i, ti in enumerate(tis):
            end_date = ti.end_date or datetime.now()
            tasks += [ti.task_id]
            color = State.color(ti.state)
            data.append({
                'x': i,
                'low': int(ti.start_date.strftime('%s')) * 1000,
                'high': int(end_date.strftime('%s')) * 1000,
                'color': color,
            })
        height = (len(tis) * 25) + 50
        session.commit()
        session.close()

        hc = {
            'chart': {
                'type': 'columnrange',
                'inverted': True,
                'height': height,
            },
            'xAxis': {'categories': tasks},
            'yAxis': {'type': 'datetime'},
            'title': {
                'text': None
            },
            'plotOptions': {
                'series': {
                    # 'borderColor': 'black',
                    'minPointLength': 5,
                },
            },
            'legend': {
                'enabled': False
            },
            'series': [{
                'data': data
            }]
        }
        return self.render(
            'admin/gantt.html',
            dag=dag,
            form=form,
            hc=json.dumps(hc, indent=4),
            height=height,
        )


admin.add_view(Flux(name='DAGs'))

# ------------------------------------------------
# Leveraging the admin for CRUD and browse on models
# ------------------------------------------------


class ModelViewOnly(ModelView):
    """
    Modifying the base ModelView class for non edit, browse only operations
    """
    named_filter_urls = True
    can_create = False
    can_edit = False
    can_delete = False
    column_display_pk = True


def filepath_formatter(view, context, model, name):
    url = url_for('flux.code', dag_id=model.dag_id)
    short_fp = model.filepath.replace(settings.BASE_FOLDER + '/dags/', '')
    link = Markup('<a href="{url}">{short_fp}</a>'.format(**locals()))
    return link


def dag_formatter(view, context, model, name):
    url = url_for('flux.tree', dag_id=model.dag_id, num_runs=45)
    link = Markup('<a href="{url}">{model.dag_id}</a>'.format(**locals()))
    return link


class DagModelView(ModelViewOnly):
    column_formatters = {
        'dag_id': dag_formatter,
        'filepath': filepath_formatter,
    }
    column_list = ('dag_id', 'task_count', 'filepath')
mv = DagModelView(models.DAG, session, name="DAGs")
admin.add_view(mv)


def dag_link(view, context, model, name):
    return model


class DagModelView(ModelViewOnly):
    column_formatters = {
        'dag_id': dag_link,
    }


class TaskModelView(ModelViewOnly):
    column_filters = ('dag_id', 'owner', 'start_date', 'end_date')
    column_formatters = {
        'dag': dag_formatter,
    }
mv = TaskModelView(
    models.BaseOperator, session, name="Tasks", category="Objects")
admin.add_view(mv)


class TaskInstanceModelView(ModelViewOnly):
    column_filters = ('dag_id', 'task_id', 'state', 'execution_date')
    column_list = (
        'state', 'dag_id', 'task_id', 'execution_date',
        'start_date', 'end_date', 'duration')
    can_delete = True
mv = TaskInstanceModelView(
    models.TaskInstance, session, name="Task Instances", category="Objects")
admin.add_view(mv)


class JobModelView(ModelViewOnly):
    column_default_sort = ('start_date', True)
mv = JobModelView(models.BaseJob, session, name="Jobs", category="Objects")
admin.add_view(mv)


mv = ModelView(models.User, session, name="Users", category="Admin")
admin.add_view(mv)


class DatabaseConnectionModelView(ModelView):
    column_exclude_list = ('login', 'password',)
    form_choices = {
        'db_type': [
            ('mysql', 'MySQL',),
            ('oracle', 'Oracle',),
        ]
    }
mv = DatabaseConnectionModelView(
    models.DatabaseConnection, session,
    name="Database Connections", category="Admin")
admin.add_view(mv)


class LogModelView(ModelViewOnly):
    column_default_sort = ('dttm', True)
    column_filters = ('dag_id', 'task_id', 'execution_date')
mv = LogModelView(models.Log, session, name="Logs", category="Admin")
admin.add_view(mv)

if __name__ == "__main__":
    logging.info("Starting the web server.")
    app.run(debug=True)
